text/trace/0168-sampling-propagation.md

```diff
@@ -0,0 +1,181 @@
+in order to build span-to-metrics pipelines when the built-in
+`ParentBased` Sampler is used.
+
+A consistent trace sampling decision is one that can be carried out at
+any node in a trace, which supports collecting partial traces.
+OpenTelemetry specifies a built-in `TraceIDRatioBased` Sampler that
+aims to accomplish this goal but was left incomplete (see
+[TODOs](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#traceidratiobased) in the specification).
+
+We propose to propagate the necessary information alongside the [W3C
+sampled flag](https://www.w3.org/TR/trace-context/#sampled-flag) using
+`tracestate` with an `otelprob` vendor tag.
+
+## Explanation
+
+Two pieces of information are needed to convey consistent head trace
+sampling probability:
+
+1. The head trace sampling probability
+2. Source of consistent sampling decisions.
+
+This proposal uses one byte of information for each of these.
+
+### Probability value
+
+To limit the cost of this extension and for statistical reasons
+documented below, we propose to limit head trace sampling probability
+to powers of two.  This limits the available head trace sampling
+probabilities to 1/2, 1/4, 1/8, and so on.  We can compactly encode
+these probabilities as small integer values using the base-2 logarithm
+of the adjusted count (i.e., inverse probability).
+
+For example, the probability value 2 corresponds with 1-in-4 sampling,
+the probability value 10 corresponds with 1-in-1024 sampling.  Using
+one byte of information we can convey sampling rates as small as 2^-255.
+
+### Random value
+
+With head trace sampling probabilities limited to powers of two, the
+amount of randomness needed per trace context is limited.  A
+consistent sampling decision is accomplished by propagating a
+geometrically distributed random variable with shape parameter `1/2`,
+requiring only two bits of randomness on average per trace.  See
+[Estimation from Partially Sampled Distributed
+Traces](https://arxiv.org/pdf/2107.07703.pdf) section 2.8 for a
+detailed explanation.
+
+Such a random variable `r` can be generated using the following
+pseudocode:
+
+```
+r := 0
+for {
+  if nextRandomBit() {
+    break // The expected value of r is 2
+  }
+  r++
+}
+```
+
+This can be computed from a stream of random bits as the number of
+leading zeros using efficient instructions on modern computer
+architectures.
+
+For example, the value 3 means there were three leading zeros and
+corresponds with being sampled at probabilities 1-in-1 through 1-in-8
+but not at probabilities 1-in-16 and smaller.  Using one byte of
+information we can convey a consistent sampling decision for sampling
+rates as small as 2^-255.
+
+### Proposed `tracestate` syntax
+
+The consistent sampling decision and head trace sampling probability
+will be propagated using four bytes of base16 content, as follows:
+
+```
+tracestate: otelprob=PPRR
```

Comment:
This does not match w3c spec (cf. https://www.w3.org/TR/trace-context/#combined-header-value), the expectation is that a vendor uses only a single key

> Example: `vendorname1=opaqueValue1,vendorname2=opaqueValue2`

We need to pick a vendor key for all of OTEL (`otel` or even just `ot`), and then decide on the encoding format for parts/attributes (which w3c does not prescribe)), for example:

```
tracestate: ot=p:PP;r:RR
tracestate: otel=pr:PPRR
```
Colons and semi-colons are allowed in values, but we need to consider if some future parts of state may need to contain (semi)colons.

Reply:
Sounds like a plan. Shall we define _first_ in the Specification the way to populate `tracestate` ourselves @yurishkuro? Happy to do that, but I'd like if that's what you foresee.

Reply:
@carlosalberto there are two independent things - how the tracestate format should look, and how SDK should interact with it. The latter is more complicated and I would say requires a prototype first, because there may be multiple touch points in the SDK that would need to mutate tracestate.

Reply:
I agree with yuri that `tracestate: otelprob=PPRR` is not the intention of the w3c spec and that we should have a single key and define our own structure within the value. `ot` or `otel` both seem fine to me. Yuri's proposed semicolon-delimited value format `key1:value1;key2:value2` seems reasonable enough to me. I would probably prefer `pr:PPRR` to `p:PP;r:RR` because you need both keys in order for the value to be useful and the first option is shorter and makes it more clear the values are tightly coupled.

Reply:
@carlosalberto has a draft on this: https://github.com/open-telemetry/opentelemetry-specification/pull/1852